nengo/builder/optimizer.py

```diff
@@ -0,0 +1,969 @@
+    on the operator type and view bases will be done with dictionaries. This
+    grouping can be done in amortized linear time and reduces the actual worst
+    case runtime of the optimization algorithm to O(g*m^2 + e) where g is the
+    number of groups and m the number of elements in a group. Moreover,
+    information about memory alignment will be used to cut the inner loop short
+    in many cases and gives a runtime much closer to linear in most cases.
+
+    Note that both `model` and `dg` will be changed by the `optimize` method.
+
+    Parameters
+    ----------
+    model : :class:`nengo.builder.builder.Model`
+        Builder output to optimize.
+    dg : dict
+        Dict of the form {a: {b, c}} where b and c depend on a specifying the
+        operator dependency graph of the model.
+    """
+
+    op_mergers = {}
+
+    enabled = BoolParam('enabled', default=True, optional=False)
+    max_passes = IntParam('max_passes', default=None, optional=True)
+
+    def __init__(self, model, dg):
+        self.enabled = RcDefault
+        self.max_passes = RcDefault
+
+        self.model = model
+        self.dg = dg
+
+        # Internal dict mapping Signals to Operators accessing those signals.
+        self._sig2op = None
+        # Internal set of Operators merged during the current optimization
+        # pass.
+        self._merged = None
+
+    @classmethod
+    def supports_merge(cls, op_type):
+        return op_type in cls.op_mergers
+
+    @classmethod
+    def can_merge_ops(cls, op1, op2):
+        return (
+            type(op1) == type(op2) and
+            type(op1) in cls.op_mergers and
+            cls.op_mergers[type(op1)].can_merge(op1, op2))
+
+    @classmethod
+    def merge_ops(cls, ops):
+        return cls.op_mergers[type(ops[0])].merge(ops)
+
+    @classmethod
+    def register_merger(cls, op_type):
+        def register(merger):
+            if op_type in cls.op_mergers:
+                warnings.warn(
+                    "Merger for operator of type {} overwritten.".format(
+                        op_type))
+            cls.op_mergers[op_type] = merger()
+            return merger
+        return register
+
+    def optimize(self):
+        """Perform the optimization."""
+        if not self.enabled:
+            return
+
+        logger.info("Running %s ...", self.__class__.__name__)
+
+        # We try first to merge operators with views only as these have a fixed
+        # order for the memory alignment whereas operators without views could
+        # be merged in a random order. Merging the views of operators will
+        # propagate requirements in the memory ordering via the other
+        # associated signals of the operator to other operators.
+        #
+        # Once no more operators with views can be merged, we try to merge
+        # operators without views and then try again merging views (because
+        # each operator merge might generate new views).
+
+        before, after = None, None
```

Comment:
I see your argument for not initializing with `None`. My argument for doing so: At that point we don't know the proper values (at least not for `after`), thus I would like to avoid to give the impression that we do. To me the initialization with `None` also seems beneficial in case a bug is introduced that prevents the proper initialization later in the loop. With the `None` initialization the code is more likely to fail early because something requires an `int` instead of silently continuing with a wrong value. Also, when printing the value for debugging purposes it would be obvious that the value is uninitialized.

If you or others still feel strongly about this after considering this, I'm willing to change to it.